### PR TITLE
Add HPK test coverage for Cosmos

### LINKED ISSUE #3666: [Cosmos] Testing: Additional sanity and coverage testing
It was noted as part of the [QueryResult pager](https://github.com/Azure/azure-sdk-for-rust/pull/3624) PR that we should have as a minimum some sanity tests for the different partition key options that are available. This issue tracks the effort of adding these, along with the effort of generally expanding the query coverage we currently have and using all types of partition keys that are available and should also have coverage.

### LINKED ISSUE #4155: Cosmos: Thorough HPK testing
Issue to track the thorough testing of HPK scenarios across different levels of prefix keys and full keys on queries.

### PR DESCRIPTION
**Pull request by simorenoh**
## Summary

Closes the end-to-end Hierarchical Partition Key (HPK) test gap tracked in #4155.

Adds a dedicated `emulator_tests::cosmos_hpk` module, extends `cosmos_feed_ranges` with HPK negative-validation tests, and adds an `in_memory_emulator_tests::cosmos_hpk_split` module covering multi-partition ("split") topologies that a single-partition emulator container cannot reproduce.

Every test is validated against both a **classic Cosmos emulator** and a **live Azure Cosmos account** — 22 passed / 0 ignored on each.

## What's covered

**Point CRUD on full hierarchical keys** — create, read, replace, upsert, delete, patch across 2- and 3-level keys.

**Key-arity rejection** — a partial key on a point op is asserted to return `400 / 1001 PartitionKeyMismatch` (status *and* sub-status). A key with more components than the container has paths returns a bare `400` with no sub-status; the test asserts exactly that, since the service does not send 1001 here.

**Queries** — full-key scope, prefix (partial-key) scopes at levels 1 and 2 plus a no-match and an anti-leak correctness guard, cross-partition fan-out over the whole container, full-key equality over `full_container`, a SQL predicate on a non-leading key path, and single-partition `ORDER BY`.

**Advanced cross-partition operators** — `DISTINCT`, aggregates, `GROUP BY`, and `OFFSET/LIMIT` are asserted to fail with the exact `400 / 1004 CrossPartitionQueryNotServable`, not merely to error. The SDK advertises no cross-partition query features, so these need a client-side pipeline it does not implement.

**Component value edge cases** — numeric, boolean, and empty-string components round-trip through create / read / query.

**Transactional batch** — happy path, plus a PK-mismatched item asserting `207 MultiStatus` with a per-op `400` and no commit.

**Feed-range negative validation** (Group C in `cosmos_feed_ranges.rs`) — empty key and too-many-components map to their client-side `CosmosStatus` values before any request is issued.

**Split topologies** (`in_memory_emulator_tests::cosmos_hpk_split`, gated on the internal `__internal_in_memory_emulator` feature) — multi-partition topology reporting, full-key single-partition targeting, deterministic distribution across partitions with per-range query filtering, and a real `split_partition` that preserves data and routing. Runs in normal CI with no live account.

## Where tests are skipped, and why

No test is ignored for an unfixed SDK gap any more. The gaps this PR originally documented — prefix filtering (#4680), cross-partition fan-out over HPK containers (#4681), and full-key equality collapsing to an empty point (#4574) — were fixed by #4729 and #4638, and the corresponding tests are now enabled and passing.

Two narrower skips remain, both environmental rather than product defects:

- The **four prefix tests** skip on the classic emulator, which predates service-accurate EPK-range HPK queries and rejects them with `400`. This was re-confirmed by temporarily lifting the skip: all four fail there, and all four pass against a live account. Prefix filtering is additionally covered deterministically in CI by `in_memory_emulator_tests::hpk`.
- **Two tests** skip on the vnext-preview emulator, which diverges from both the classic emulator and live accounts. Tracked in #4742.

## Notes

Test-only change; no library code is modified. No CHANGELOG entry, per the convention of leaving an Unreleased section untouched when no public API changes. `cargo fmt` and `cargo clippy` are clean.

Relates to #3666.

### CODE REVIEW

File: sdk/cosmos/azure_data_cosmos/tests/emulator_tests/cosmos_hpk.rs [hunk 1]
```diff
@@ -0,0 +1,780 @@
...
+// Licensed under the MIT License.
+
+//! End-to-end emulator coverage for Hierarchical Partition Keys (HPK).
+//!
+//! These tests exercise the public `azure_data_cosmos` surface against
+//! MultiHash (hierarchical) containers: point CRUD with full keys, partial-key
+//! and too-many-component rejection, full-key and prefix queries, cross-partition
+//! queries, component value edge cases (null / numeric / bool / empty string),
+//! and transactional batch. Single-hash feed-range negatives live alongside the
+//! positive feed-range tests in `cosmos_feed_ranges.rs`.
```
[the remark this answers is not in the document]
Comment (simorenoh) [reply sdk/cosmos/azure_data_cosmos/tests/emulator_tests/cosmos_hpk.rs]: Fixed in d97e6677e6 — retitled these as HPK feed-range negatives (C1/C2 are multi-path; only C3 is single-hash).

File: sdk/cosmos/azure_data_cosmos/tests/emulator_tests/cosmos_hpk.rs [hunk 2]
```diff
@@ -0,0 +1,780 @@
...
+///
+/// Prefix (partial-key) HPK queries are issued through
+/// [`FeedScope::partition`] with fewer components than the container has paths.
+/// Despite its doc comment recommending a full key, `FeedScope::partition`
+/// builds a logical-partition feed range whose effective partition key is
+/// computed from the supplied prefix, which the gateway expands to the matching
+/// sub-range. Resolving the prefix through
+/// [`ContainerClient::feed_range_from_partition_key`] instead would return the
+/// *physical* partition range (the whole partition) and therefore would not
+/// filter by the prefix.
```
[the remark this answers is not in the document]
Comment (simorenoh) [reply sdk/cosmos/azure_data_cosmos/tests/emulator_tests/cosmos_hpk.rs]: Good catch — fixed in d97e6677e6. The doc now states prefix queries are NOT filtered today (single-point EPK via `compute`, not `compute_range`), matching the `#[ignore]` reasons and #4680.

File: sdk/cosmos/azure_data_cosmos/tests/emulator_tests/cosmos_hpk.rs [hunk 3]
```diff
@@ -0,0 +1,780 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+//! End-to-end emulator coverage for Hierarchical Partition Keys (HPK).
+//!
+//! These tests exercise the public `azure_data_cosmos` surface against
+//! MultiHash (hierarchical) containers: point CRUD with full keys, partial-key
+//! and too-many-component rejection, full-key and prefix queries, cross-partition
+//! queries, component value edge cases (null / numeric / bool / empty string),
```
[the remark this answers is not in the document]
Comment (simorenoh) [reply sdk/cosmos/azure_data_cosmos/tests/emulator_tests/cosmos_hpk.rs]: Fixed in d97e6677e6 — removed `null` from the module doc's edge-case list.

File: sdk/cosmos/azure_data_cosmos/CHANGELOG.md [hunk 4]
```diff
@@ -14,6 +14,8 @@
 
 ### Other Changes
 
+- Added an `emulator_tests::cosmos_hpk` integration-test module and extended `cosmos_feed_ranges` with hierarchical-partition-key (HPK) negative validation tests, covering item CRUD, patch, full-key/prefix/cross-partition queries, component value edge cases (numeric/bool/null/empty), transactional batch, and feed-range validation against the Cosmos emulator. ([#4155](https://github.com/Azure/azure-sdk-for-rust/issues/4155))
```
[the remark this answers is not in the document]
Comment (simorenoh) [reply sdk/cosmos/azure_data_cosmos/CHANGELOG.md]: Fixed in d97e6677e6 — dropped `null` from the CHANGELOG entry; only numeric/bool/empty-string component tests exist.

File: sdk/cosmos/azure_data_cosmos/tests/emulator_tests/cosmos_hpk.rs [hunk 5]
```diff
@@ -0,0 +1,780 @@
...
+        .create_container(db_client, properties, None)
+        .await?;
+    Ok(container)
+}
+
+/// Drains a query to completion via its page iterator, returning every item.
+///
+/// All emulator query tests in this suite use the page iterator (`into_pages`)
+/// rather than the item-level `TryStream`, because cross-partition fan-out is
+/// only driven correctly through the paged interface.
```
[the remark this answers is not in the document]
Comment (simorenoh) [reply sdk/cosmos/azure_data_cosmos/tests/emulator_tests/cosmos_hpk.rs]: Fixed in d97e6677e6 — scoped the claim: only the cross-partition fan-out cases require `into_pages`; single-partition tests may use either interface.

**Comment by simorenoh** [comment]
@microsoft-github-policy-service rerun

**Review by analogrelay — APPROVED** [review]
IF the weekly test failures are flakiness, I'm OK with you merging while they're red. Please make a comment showing that for posterity's sake though :). We can open a separate issue for looking in to the flakiness.

File: sdk/cosmos/azure_data_cosmos/CHANGELOG.md [hunk 6]
```diff
@@ -14,6 +14,10 @@
...
+- Added an `in_memory_emulator_tests::cosmos_hpk_split` module (gated on the internal `__internal_in_memory_emulator` feature) that exercises HPK over a multi–physical-partition ("split") topology using the in-memory emulator: multi-partition topology reporting, full-key single-partition targeting, deterministic dataset distribution across partitions with cross-partition query fan-out, and a real `split_partition` that preserves data and routing. This covers the Class C split scenarios that a single-partition emulator container cannot reproduce, with no live account required. ([#4155](https://github.com/Azure/azure-sdk-for-rust/issues/4155))
+
+
```
[the remark this answers is not in the document]
Comment (simorenoh) [reply sdk/cosmos/azure_data_cosmos/CHANGELOG.md]: Applied in e6ab556139. Agreed - this PR is test-only, and the repo convention is to leave an Unreleased section alone when no public API changes, so both entries are removed.

File: sdk/cosmos/azure_data_cosmos/tests/in_memory_emulator_tests/cosmos_hpk_split.rs [hunk 7]
```diff
@@ -0,0 +1,402 @@
...
+/// c04 — A multi-country dataset spreads deterministically across both physical
+/// partitions (proven via the SDK's `feed_range_from_partition_key` routing), and
+/// a cross-partition `SELECT * FROM c` fans out to both partitions and returns the
+/// whole dataset.
+///
+/// Note: distribution is asserted through feed-range *resolution* rather than
+/// `FeedScope::range`-scoped queries, because the in-memory emulator's query engine
+/// does not filter documents by EPK sub-range (a range-scoped `SELECT *` returns the
+/// full container). Range-scoped query filtering is validated against the real
+/// emulator in `cosmos_feed_ranges.rs` / `cosmos_hpk.rs`.
```
[the remark this answers is not in the document]
Comment (simorenoh) [reply sdk/cosmos/azure_data_cosmos/tests/in_memory_emulator_tests/cosmos_hpk_split.rs]: Same as above - fixed in 9d75f30579; the outdated EPK-filtering claim is gone and the distribution test now checks per-range results.

File: sdk/cosmos/azure_data_cosmos/tests/emulator_tests/cosmos_hpk.rs [hunk 8]
```diff
@@ -0,0 +1,1036 @@
...
+
+/// B9: cross-partition advanced query operators (`DISTINCT`, aggregates,
+/// `GROUP BY`, `OFFSET/LIMIT`) over a hierarchical container are not servable.
+///
+/// The Rust SDK advertises no cross-partition query features
+/// (`SUPPORTED_QUERY_FEATURES=""`), so any full-container query that needs a
+/// client-side pipeline is rejected. On hierarchical containers this is
+/// compounded by the cross-partition fan-out defect (#4681), so these fail
+/// regardless of operator. Each case is expected to error (400 BadRequest /
+/// 1004 CrossPartitionQueryNotServable).
```
[the remark this answers is not in the document]
Comment (simorenoh) [reply sdk/cosmos/azure_data_cosmos/tests/emulator_tests/cosmos_hpk.rs]: Applied in e6ab556139. Now attributes the failures solely to the missing cross-partition operator pipeline, and notes HPK fan-out works since #4729.

File: sdk/cosmos/azure_data_cosmos/tests/in_memory_emulator_tests/cosmos_hpk_split.rs [hunk 9]
```diff
@@ -0,0 +1,402 @@
...
+//! * **Topology** — a MultiHash `/country/state/city` container provisioned with 2
+//!   physical partitions reports exactly 2 feed ranges to the client.
+//! * **Full-key targeting** — a complete 3-level key always resolves to exactly one
+//!   physical partition, even across a split topology.
+//! * **Distribution / routing correctness** — a multi-country dataset spreads
+//!   deterministically across both partitions (proven via the SDK's
+//!   `feed_range_from_partition_key` routing), and a cross-partition query fans out
+//!   to both partitions and returns the whole dataset. The in-memory emulator's
+//!   query engine does not honor `FeedScope::range` (EPK sub-range) filtering, so
+//!   range-scoped result filtering is validated against the real emulator elsewhere.
```
[the remark this answers is not in the document]
Comment (simorenoh) [reply sdk/cosmos/azure_data_cosmos/tests/in_memory_emulator_tests/cosmos_hpk_split.rs]: Already addressed in 9d75f30579 (pushed before this review). c04 now asserts per-range filtering directly: each range-scoped query returns exactly the count routing predicted, every returned document routes back to the range that served it, and the per-range results partition the dataset exactly (verified 3+5=8, no loss or duplication).

File: sdk/cosmos/azure_data_cosmos/tests/emulator_tests/cosmos_hpk.rs [hunk 10]
```diff
@@ -0,0 +1,1036 @@
...
+                .read_item(prefix, "la-1", None)
+                .await
+                .expect_err("partial-key point read should fail");
+            // The gateway rejects a partial logical key on a point operation
+            // with 400 BadRequest (sub-status 1001 PartitionKeyMismatch).
+            assert_eq!(
+                err.status().status_code(),
+                StatusCode::BadRequest,
+                "partial-key point read should be rejected with 400 BadRequest"
+            );
```
[the remark this answers is not in the document]
Comment (simorenoh) [reply sdk/cosmos/azure_data_cosmos/tests/emulator_tests/cosmos_hpk.rs:419]: Partly applied in e6ab556139, and the verification turned up something worth flagging. The partial-key case (line 420) does return `400 / 1001 PartitionKeyMismatch`, so it now asserts the sub-status. The too-many-components case (line 450) does **not** - the service returns a bare 400 with no sub-status at all. The comment claiming 1001 there was wrong, so I corrected the comment rather than the assertion; adding the suggested sub-status assert makes the test fail.

File: sdk/cosmos/azure_data_cosmos/tests/emulator_tests/cosmos_hpk.rs [hunk 11]
```diff
@@ -0,0 +1,1036 @@
...
+/// with fewer components than the container has paths. NOTE: this does **not**
+/// filter by the prefix today. `FeedScope::partition` lowers to
+/// `FeedRange::for_partition`, which computes a single-*point* effective
+/// partition key (`EffectivePartitionKey::compute`, not `compute_range`), so the
+/// query routes to one physical partition and scans it in full, returning every
+/// item in that partition regardless of the prefix. The
+/// [`ContainerClient::feed_range_from_partition_key`] route is no better — it
+/// resolves the *physical* partition range (the whole partition). There is no
+/// supported filtered-prefix query path yet; see the `#[ignore]`d prefix tests
+/// below and issue #4680.
```
[the remark this answers is not in the document]
Comment (simorenoh) [reply sdk/cosmos/azure_data_cosmos/tests/emulator_tests/cosmos_hpk.rs]: Applied in e6ab556139. The doc now describes the post-#4729 `compute_range` / `EffectivePartitionKeyRange` behavior and points at the classic-emulator runtime skip as the reason those tests skip locally.

File: sdk/cosmos/azure_data_cosmos/tests/in_memory_emulator_tests/cosmos_hpk_split.rs [hunk 12]
```diff
@@ -0,0 +1,402 @@
...
+//!
+//! Two Class C behaviors are intentionally covered elsewhere rather than here:
+//! * **Cross-partition prefix fan-out** — a partition-key *prefix* whose EPK band
+//!   spans multiple physical partitions. The in-memory emulator only ever splits at
+//!   the geometric EPK midpoint (`compute_epk_midpoint`), never inside a single
+//!   prefix's hash band, so a shared-prefix dataset always lands wholly on one side
+//!   and a prefix resolves to exactly one range (verified below in
+//!   [`hpk_split_prefix_resolves_to_single_partition`]). Real prefix-query behavior
+//!   against a genuine emulator is covered by the `hpk_query_prefix_*` tests in
+//!   `cosmos_hpk.rs` (tracked by #4680).
```
[the remark this answers is not in the document]
Comment (simorenoh) [reply sdk/cosmos/azure_data_cosmos/tests/in_memory_emulator_tests/cosmos_hpk_split.rs]: Applied in e6ab556139. Both cross-references are updated: #4680 is no longer presented as an open gap, and the full-key-equality bullet no longer describes an ignored test.

File: sdk/cosmos/azure_data_cosmos/tests/emulator_tests/cosmos_hpk.rs [hunk 13]
```diff
@@ -0,0 +1,1036 @@
...
+/// B10: a full-key equality predicate over the whole container should route to
+/// the single owning partition (closed point) and return the matching docs.
+///
+/// Currently ignored: the full-key equality/`IN` collapse-to-point fix lives in
+/// PR #4638 (issue #4574), which is not yet merged. Until then this predicate
+/// shape either panics the query worker or is rejected. Re-enable once #4638
+/// lands.
+#[tokio::test]
+#[ignore = "GAP (#4574/#4638): full-key equality over full_container collapses to an empty point and \
+            is not servable until PR #4638 merges. Re-enable once the equality/IN point-routing fix lands."]
```
[the remark this answers is not in the document]
Comment (simorenoh) [reply sdk/cosmos/azure_data_cosmos/tests/emulator_tests/cosmos_hpk.rs]: Applied in e6ab556139. Confirmed empirically: with #4638 merged, this test passes unmodified on both the classic emulator and a live account, so it now carries the ordinary emulator-category gate. 22 passed / 0 ignored on both targets.

File: sdk/cosmos/azure_data_cosmos/tests/emulator_tests/cosmos_hpk.rs [hunk 14]
```diff
@@ -0,0 +1,1036 @@
...
+/// B7: a SQL filter on a *non-leading* partition-key path, scoped to a fully
+/// routed (single logical partition) 3-level key, is served by the backend.
+///
+/// This deliberately scopes to a **full** key rather than a prefix: prefix
+/// scoping (`FeedScope::partition(("USA", "CA"))`) lowers to a single-point EPK
+/// and is not reliably servable across emulator versions — that behavior is the
+/// #4680 gap, covered by the ignored `hpk_query_prefix_*` tests. Here we prove
+/// the complementary, supported case: within a routed partition the server
+/// accepts and applies a SQL predicate that references a non-leading key path
+/// (`/city`).
```
[the remark this answers is not in the document]
Comment (simorenoh) [reply sdk/cosmos/azure_data_cosmos/tests/emulator_tests/cosmos_hpk.rs]: Applied in e6ab556139. Rationale is now about testing the full-key/non-leading-path case, with no stale #4680 claim.

File: sdk/cosmos/azure_data_cosmos/tests/emulator_tests/cosmos_hpk.rs [hunk 15]
```diff
@@ -0,0 +1,1036 @@
...
+                let result = collect_query::<serde_json::Value>(
+                    &container,
+                    query,
+                    FeedScope::full_container(),
+                )
+                .await;
+                assert!(
+                    result.is_err(),
+                    "cross-partition query should not be servable on an HPK container: {query}"
+                );
```
[the remark this answers is not in the document]
Comment (simorenoh) [reply sdk/cosmos/azure_data_cosmos/tests/emulator_tests/cosmos_hpk.rs:848]: Applied in e6ab556139. All four cases now assert `CosmosStatus::CROSS_PARTITION_QUERY_NOT_SERVABLE` via a downcast, matching the `cosmos_query.rs` pattern. Verified against the emulator (400/1004 on all four) and a live account.

**Review by tvaron3 — APPROVED** [review]

**Review by analogrelay — APPROVED** [review]

**Review by analogrelay — APPROVED** [review]
